include/thread_pool/rouser.hpp

```diff
@@ -0,0 +1,150 @@
+    * active, or in an otherwise undefined state.
+    */
+    Rouser& operator=(Rouser&& rhs) noexcept;
+
+    /**
+    * @brief start Create the executing thread and start tasks execution.
+    * @param workers A pointer to the vector containing sibling workers for performing round robin work stealing.
+    * @param idle_workers A pointer to the slotted bag containing all idle workers.
+    * @param num_busy_waiters A pointer to the atomic busy waiter counter.
+    * @note The parameters passed into this function generally relate to the global thread pool state.
+    */
+    template <typename Task, template<typename> class Queue>
+    void start(std::vector<std::unique_ptr<Worker<Task, Queue>>>& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters);
+
+    /**
+    * @brief stop Stop all worker's thread and stealing activity.
+    * Waits until the executing thread becomes finished.
+    */
+    void stop();
+
+private:
+
+    /**
+    * @brief threadFunc Executing thread function.
+    * @param workers A pointer to the vector containing sibling workers for performing round robin work stealing.
+    * @param idle_workers A pointer to the slotted bag containing all idle workers.
+    * @param num_busy_waiters A pointer to the atomic busy waiter counter.
+    */
+    template <typename Task, template<typename> class Queue>
+    void threadFunc(std::vector<std::unique_ptr<Worker<Task, Queue>>>& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters);
+
+    std::atomic<bool> m_running_flag;
+    std::atomic<bool> m_started_flag;
+    std::thread m_thread;
+    std::chrono::microseconds m_rouse_period;
+};
+
+inline Rouser::Rouser(std::chrono::microseconds rouse_period)
+    : m_running_flag(false)
+    , m_started_flag(false)
+    , m_rouse_period(std::move(rouse_period))
+{
+}
+
+inline Rouser::Rouser(Rouser&& rhs) noexcept
+{
+    *this = std::move(rhs);
+}
+
+inline Rouser& Rouser::operator=(Rouser&& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        m_running_flag = rhs.m_running_flag.load();
+        m_started_flag = rhs.m_started_flag.load();
+        m_thread = std::move(rhs.m_thread);
+        m_rouse_period = std::move(rhs.m_rouse_period);
+    }
+
+    return *this;
+}
+
+template <typename Task, template<typename> class Queue>
+inline void Rouser::start(std::vector<std::unique_ptr<Worker<Task, Queue>>>& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters)
+{
+    if (m_started_flag.exchange(true, std::memory_order_acq_rel))
+        throw std::runtime_error("The Rouser has already been started.");
+
+    m_running_flag.store(true, std::memory_order_release);
+    m_thread = std::thread(&Rouser::threadFunc<Task, Queue>, this, std::ref(workers), std::ref(idle_workers), std::ref(num_busy_waiters));
+}
+
+inline void Rouser::stop()
+{
+    if (m_running_flag.exchange(false, std::memory_order_acq_rel))
```

Comment:
You probably need a symmetrical exception here. in start function you throw exception when it is called on already started object. The opposite should be done here.

Reply:
Implemented!